fix: semaphore counting

RianFuro opened the pull request:
Fix counting semaphore around blocking threshold.
Add some test cases to cover basic behavior around blocking threshold.

cc @oberblastmeister

oberblastmeister commented:
I didn't have the chance to test the semaphore so thanks for the help!

oberblastmeister commented on tests/plenary/async_lib/semaphore_spec.lua:
```diff
@@ -12,6 +12,40 @@ a.describe('semaphore', function()
+
+    local blocking_future = blocking()
+    permit:forget()
+    a.util.block_on(blocking_future)
```
why not `a.run`?

RianFuro replied:
gap in knowledge :D can I be sure that the future is run to completion with `a.run`? blocking just seemed like the intuitive thing to do.

oberblastmeister replied:
actually my bad `await` would actually be the right thing to use here. You are right, `a.run` will not necessarily run it to completion within the function but it probably will in this case because we are not doing any real async such as io, we are just using coroutine abstractions. `await` will always run to completion. The test is automatically blocked on also.

oberblastmeister commented:
You should also probably add a test where the the permit is forgotten while someone is waiting on it, not just before someone is waiting for it so that they automatically get it.

RianFuro commented:
> You should also probably add a test where the the permit is forgotten while someone is waiting on it, not just before someone is waiting for it so that they automatically get it.

That's what I attempted with the last test actually. Good catch. I updated the test and added another one that just tests sequential acquisition.

RianFuro commented on tests/plenary/async_lib/semaphore_spec.lua:
```diff
@@ -12,6 +12,47 @@ a.describe('semaphore', function()
+      completed = true
+    end)
+
+    a.run(blocking())
```
I used `a.run` this time to avoid blocking, since I need to release the permit after. `run` should at least immediately unwind the coroutine until the next yield, so the acquire definitely blocks at that point (confirmed with print logs). I'm not sure how stable that test is though; never failed after a couple runs anyway.

Any suggestion how to sequence that better?

RianFuro replied:
Actually, I believe calling forget should then unwind the rest of the coroutine synchronously since it calls the stored continuation... right?

oberblastmeister replied:
Yes `forget` will unwind the coroutine because it stored the callback

oberblastmeister replied:
The test is very stable because we are not doing any io so everything will always run the same way.
